enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,314 @@
+
+### Suggested design
+
+To support early clustering requirements Keepalived will continue running as static pods through MCO, additionally, a new dnsmasq service (also deployed through MCO) will run in each node while HAProxy, CoreDNS-MDNS and MDNS-publisher will run as daemonset pods through the new operator.
+
+The following [diagram](https://docs.google.com/drawings/d/1a8K16vcaBB6SIF0bN8DaTj4ltjBxVsVNjJN5NCd-U0A/edit?usp=sharing) reflects the suggested changes.
+
+#### Self hosted DNS services
+- The self hosted DNS functionality should be composed of the following components in each node:
+    - A NetworkManager dispatcher script used to configure resolv.conf to point at the node's public IP address, see [self hosted DNS design](https://github.com/openshift/installer/blob/f6bb83784a835e1f4955ef41916793c0948fd4a1/docs/design/baremetal/networking-infrastructure.md#internal-dns) for more details.
+    - Systemd service running dnsmasq
+    - CoreDNS-mDNS pod
+- The CHNSO will be responsible only for the deployment of the CoreDNS pods while the NetworkManager dispatcher script and dnsmasq Systemd service will run through MCO.
+- To avoid port conflict, Dnsmasq service should listen on DNS well-known port (53) and CoreDNS-MDNS should listen on other port (e.g: 5353). 
+- The new Dnsmasq service should be responsible to resolve 'api-int' (to API-VIP value provided in install-config) hostname when CoreDNS-MDNS pod isn't running on the node.
+- CoreDNS-MDNS should register himself as an upstream server in dnsmasq [via dbus interface](https://github.com/imp/dnsmasq/blob/master/dbus/DBus-interface#L51) for ClusterName.DomainName
+ 
+#### Self hosted API Loadbalancer
+- The self hosted API loadbalancer will run similarly to the current mode with one difference: HAProxy pods will run by CHSNO as DaemonSet pods instead of the static pods.
+- API should be available even when HAProxy pods don't run.
+- Additional functionality aspects of the self-hosted Loadbalancer maybe further analyzed upon need in a separate enhancement. 
+
+#### Master node shutdown
+
+In order not to break the [graceful termination process of kube-api](https://github.com/openshift/installer/blob/master/docs/dev/kube-apiserver-health-check.md) the pods responsible for API reachability should be deleted in the following order: kube-api, Keepalived, and HAProxy.
+
+Using [K8S graceful node shutdown in kubelet](https://github.com/kubernetes/kubernetes/pull/96129) it should be possible to delete the pods in the order described above.
+
+A detailed design of pods priority will be defined after graceful node shutdown in kubelet feature will be integrated in Openshift.
+
+Question: Isn't it possible to force deletion of kube-api, Keepalived, and HAProxy in the desired order by implementing smart delay code in SIGTERM handler in each one of the pods?
```

Comment:
that's a good question. Both keepalived and haproxy sigterm code could watch the kube-apiserver pod via shared pid namespace, or some other means.

Reply:
q: can and would you patch keepalived and haproxy binaries?

Reply:
That's should doable. 
Both HAProxy and Keepalived (see [1] and [2]) containers run as bash scripts ( to support Rx and process commands from monitor sidecar container ) there shouldn't be any problem to add SIGTERM handler to both of them.

[1] https://github.com/openshift/machine-config-operator/blob/master/templates/master/00-master/on-prem/files/haproxy.yaml#L38-#L91
[2]  https://github.com/openshift/machine-config-operator/blob/master/templates/common/on-prem/files/keepalived.yaml#L69#L102